File: example/multiclassification/multiclassification.go
package main

import (
	"fmt"
	"log"
	"path"
	"path/filepath"
	"runtime"

	"github.com/mirecl/catboost-cgo/catboost"
)

func main() {
	_, fileName, _, _ := runtime.Caller(0)
	modelPath := path.Join(filepath.Dir(fileName), "multiclassification.cbm")

	// Initialize CatBoostClassifier
	model, err := catboost.LoadFullModelFromFile(modelPath)
	if err != nil {
		log.Fatalln(err)
	}

	// Initialize data
	floats := [][]float32{{1996, 197}, {1968, 37}, {2002, 77}, {1948, 59}}
	cats := [][]string{{"winter"}, {"winter"}, {"summer"}, {"summer"}}

	// Get batch predicted RawFormulaVal
	preds, err := model.Predict(floats, cats)
	if err != nil {
		log.Fatalln(err)
	}

	predsMulti := model.Transform(preds)
	fmt.Printf("Preds `RawFormulaVal`: %.8f\n", predsMulti)

	// Get single predicted RawFormulaVal
	pred, err := model.PredictSingle(floats[0], cats[0])
	if err != nil {
		log.Fatalln(err)
	}
	fmt.Printf("Pred `RawFormulaVal`: %.8f\n", pred)

	// Get batch predicted probabilities for each class
	model.SetPredictionType(catboost.Probablity)
	preds, err = model.Predict(floats, cats)
	if err != nil {
		log.Fatalln(err)
	}

	predsMulti = model.Transform(preds)
	fmt.Printf("Preds `Probability`: %.8f\n", predsMulti)

	// Get single predicted probabilities for each class
	pred, err = model.PredictSingle(floats[0], cats[0])
	if err != nil {
		log.Fatalln(err)
	}
	fmt.Printf("Pred `Probability`: %.8f\n", pred)

	// Get batch predicted classes
	model.SetPredictionType(catboost.Class)
	preds, err = model.Predict(floats, cats)
	if err != nil {
		log.Fatalln(err)
	}

	predsMulti = model.Transform(preds)
	fmt.Printf("Preds `Class`: %.0f\n", predsMulti)

	// Get single predicted classes
	pred, err = model.PredictSingle(floats[0], cats[0])
	if err != nil {
		log.Fatalln(err)
	}
	fmt.Printf("Pred `Class`: %.0f\n", pred)
}

File: example/multiclassification/multiclassification.py
# https://catboost.ai/en/docs/concepts/python-usages-examples#multiclassification
from catboost import Pool, CatBoostClassifier
import pathlib

path = pathlib.Path(__file__).parent.resolve()

train_data = [
    ["summer", 1924, 44],
    ["summer", 1932, 37],
    ["winter", 1980, 37],
    ["summer", 2012, 204],
]
eval_data = [
    ["winter", 1996, 197],
    ["winter", 1968, 37],
    ["summer", 2002, 77],
    ["summer", 1948, 59],
]

cat_features = [0]

train_label = ["France", "USA", "USA", "UK"]
eval_label = ["USA", "France", "USA", "UK"]

train_dataset = Pool(data=train_data, label=train_label, cat_features=cat_features)
eval_dataset = Pool(data=eval_data, label=eval_label, cat_features=cat_features)

# Initialize CatBoostClassifier
model = CatBoostClassifier(
    iterations=10, learning_rate=1, depth=2, loss_function="MultiClass", silent=True
)

# Fit model
model.fit(train_dataset)

# Get predicted RawFormulaVal
preds_raw = model.predict(eval_dataset, prediction_type="RawFormulaVal")
print(f"Preds `RawFormulaVal`: {preds_raw}")

# Get predicted probabilities for each class
preds_proba = model.predict_proba(eval_dataset)
print(f"Preds `Probability`: {preds_proba}")

# Get predicted classes
preds_class = model.predict(eval_dataset)
print(f"Preds `Class`: {preds_class}")

# Save model
model.save_model(f"{path}/multiclassification.cbm")
